### PR TITLE
buntime funtime

### PR DESCRIPTION
**Pull request by tim-smart**

### CODE REVIEW

File: src/browser.ts [hunk 1]
```diff
@@ -2,13 +2,8 @@ import { $ } from "bun";
...
-    return Effect.tryPromise({
-      try: async () => $`open ${uri.toString()}`,
-      catch: (error) =>
-        new Error(
-          `An error occured while trying to open the browser: ${error}`,
-        ),
-    });
-  },
+  open: (uri: URL) =>
+    Effect.promise(() => $`open ${uri.toString()}`).pipe(
```
Comment (tim-smart) [inline src/browser.ts:6]: If you don't care about the error you can use .promise

File: src/message-pubsub.ts [hunk 2]
```diff
@@ -49,67 +49,55 @@ export class MessagePubSub extends Context.Tag("message-pubsub")<
...
-
-                  if (predicate(message)) {
-                    yield* Queue.offer(queue, message);
-                  }
-
-                  yield* Effect.logInfo(`consumer for ${messageType} ending`);
-                  return yield* Effect.void;
-                }),
-              ),
+            const subscription = yield* PubSub.subscribe(pubsub);
```
Comment (tim-smart) [inline src/message-pubsub.ts:78]: When subscribing to a pub sub, you first acquire the queue, then continuously pull from it.

File: src/spotify-service.ts [hunk 3]
```diff
@@ -1,20 +1,35 @@
...
+}> {}
+
+const make = Effect.gen(function* () {
+  const config = yield* SpotifyConfigService;
+
+  const client = SpotifyApi.withAccessToken(
+    config.clientId,
+    config.accessToken
+  );
+  const use = <A>(f: (client: SpotifyApi) => Promise<A>) =>
```
Comment (tim-smart) [inline src/spotify-service.ts:17]: One way to wrap Promise apis, is to create a client wrapper that adds a custom error type.

File: src/spotify-service.ts [hunk 4]
```diff
@@ -1,67 +1,78 @@
...
 
-export const SpotifyService = Layer.effectDiscard(
-  Effect.scoped(
-    Effect.gen(function* () {
-      yield* Effect.logInfo("starting spotify service");
-      const api = yield* SpotifyApiClient;
-      const pubsub = yield* MessagePubSub;
-
-      const dequeue = yield* pubsub.subscribeTo("CurrentlyPlayingRequest");
+export const SpotifyService = Layer.scopedDiscard(
```
Comment (tim-smart) [inline src/spotify-service.ts:35]: You can use Layer.scopedDiscard instead of Effect.scoped here

File: src/spotify-service.ts [hunk 5]
```diff
@@ -1,67 +1,78 @@
...
-    }),
-  ).pipe(Effect.provide(Layer.mergeAll(SpotifyApiClient.Live))),
-);
+          yield* Effect.logInfo("publishing currently playing");
+          yield* pubsub.publish(Message.CurrentlyPlaying({ song: item }));
+        })
+      )
+    );
+  })
+).pipe(Layer.provide(SpotifyApiClient.Live), Layer.provide(MessagePubSub.Live));
```
Comment (tim-smart) [inline src/spotify-service.ts:68]: Layer.provide > Effect.provide

File: src/spotify-service.ts [hunk 6]
```diff
@@ -1,67 +1,78 @@
...
 
 export function requestAccessToken(code: string) {
   return Effect.gen(function* () {
     const config = yield* SpotifyConfigService;
-    const authorizationHeader = `Basic ${Encoding.encodeBase64(`${config.clientId}:${Secret.value(config.clientSecret)}`)}`;
+    const authorizationHeader = `Basic ${Encoding.encodeBase64(
+      `${config.clientId}:${Secret.value(config.clientSecret)}`
+    )}`;
 
     // TODO: Refactor to platform HttpClient
```
Comment (tim-smart) [inline src/spotify-service.ts:77]: https://github.com/tim-smart/effect-twitch-integrations/blob/platform-htp/src/spotify-service.ts#L83

File: src/message-pubsub.ts [hunk 7]
```diff
@@ -49,67 +49,55 @@ export class MessagePubSub extends Context.Tag("message-pubsub")<
...
+            yield* Effect.gen(function* () {
+              const message = yield* subscription.take;
+              yield* Effect.logInfo(`consumer received ${message}`);
+              if (predicate(message)) {
+                yield* Queue.offer(queue, message);
+              }
+            }).pipe(Effect.forever, Effect.forkScoped);
+
+            yield* Effect.addFinalizer(() =>
+              Effect.logInfo(`consumer for ${messageType} ending`)
```
Comment (dmmulroy) [inline src/message-pubsub.ts:95]: Ah clever I really like this

File: src/message-pubsub.ts (hunk 2 above)
Comment (dmmulroy) [reply src/message-pubsub.ts:78]: Ah so previously I was continuously resubscribing then?

File: src/spotify-config-service.ts [hunk 8]
```diff
@@ -1,41 +1,32 @@
...
-const accessToken: AccessToken = AccessTokenJson as unknown as AccessToken;
+const accessToken: AccessToken = undefined as unknown as AccessToken;
 
-export type ISpotifyConfigService = Readonly<{
-  accessToken: AccessToken;
-  clientId: string;
-  clientSecret: Secret.Secret;
-  port: number;
-  redirectServerPath: string;
-}>;
```
Comment (dmmulroy) [inline src/spotify-config-service.ts:14]: blasphemy 😆

File: src/spotify-service.ts (hunk 3 above)
Comment (dmmulroy) [reply src/spotify-service.ts:17]: Nice, I dig it.

File: src/spotify-service.ts (hunk 4 above)
Comment (dmmulroy) [reply src/spotify-service.ts:35]: TIL

File: src/spotify-service.ts (hunk 5 above)
Comment (dmmulroy) [reply src/spotify-service.ts:68]: Could you explain why?

File: src/spotify-service.ts (hunk 6 above)
Comment (dmmulroy) [reply src/spotify-service.ts:77]: This is slick 🔥

File: src/twitch-service.ts [hunk 9]
```diff
@@ -59,120 +80,111 @@ export class TwitchEventSubClient extends Context.Tag("twitch-eventsub-client")<
...
-        Effect.succeed(new EventSubWsListener({ apiClient })),
-        (eventsub) => {
-          return Effect.gen(function* () {
+        Effect.sync(() => new EventSubWsListener({ apiClient: api.client })),
+        (eventsub) =>
+          Effect.gen(function* () {
             yield* Effect.logInfo("event sub stopping");
             eventsub.stop();
-
             return Effect.void;
```
Comment (dmmulroy) [inline src/twitch-service.ts:91]: This can be omitted right?

File: src/twitch-service.ts [hunk 10]
```diff
@@ -59,120 +80,111 @@ export class TwitchEventSubClient extends Context.Tag("twitch-eventsub-client")<
...
+
+    yield* Effect.gen(function* () {
+      yield* Effect.logInfo("waiting for song");
+      const { song } = yield* Queue.take(dequeue);
+      yield* Effect.logInfo("received song");
+      yield* api.use((_) =>
+        _.chat.sendChatMessage(
+          config.broadcasterId,
+          `The currently playing song is ${song.name}`
+        )
```
Comment (dmmulroy) [inline src/twitch-service.ts:186]: I was wondering where `use` was going to show up, this is a great little addition

**Review by dmmulroy — APPROVED** [review]

File: src/twitch-service.ts (hunk 9 above)
Comment (tim-smart) [reply src/twitch-service.ts:91]: Yes, it's not being yielded either 🙈

File: src/spotify-service.ts (hunk 5 above)
Comment (tim-smart) [reply src/spotify-service.ts:68]: When you compose Layer's it ensures the same Layer will never be built more than once.

Effect.provide will build the Layer every time.

**Pull request merged**